Design.md

```diff
@@ -0,0 +1,107 @@
+
+All fragments are IMMUTABLE. Any mutation is handled at the CloudServer level.
+
+## Key generation scheme
+
+List of detailed requirements:
+Requirement | Description
+----------- | -------------
+Version | A perfect solution often does not exists. And will not stay that way. Generated keys MUST start with a version, so that we can easily change the format.
+Relate easily to parent CloudServer objectkey | Would make life easy to be able to relate a fragment key with its parent, main object.
+Small and self describing | Smallest keys possible, obviously. But without relying too much on implicit information. Generated keys should be as self describing as possible.
+Embed replication policy | Describe what code is used. Typically RS2+1 or RSk+m or even replication factor for very small object (if we decide to go that way). We could then add new code types easily.
+Split ready | A splitted key should be obvious to spot. The generated key length should NOT depend on number of chunk.
+Fragment name | Resulting fragment key stored on hyperdrive must contain the parent, S3 object name. Or at least a prefix if too long. Resulting key must also be unique hyperdrive-wise.
+Fragment location | Each fragment must enable the hdclient to select the correct hyperdrive
+Fragment type |Each fragment should be tagged as data or coding (reading would then first prefer data parts since RS is systematic)
+
+Since we can specify which key to use on an hyperdrive, we could define a new key scheme to encode fragments in a unified key scheme. The idea is to minimize the amount of metadata we have to store with every object in the MongoDB database that can be used to locate the fragments. Note that we plan to use a 2+1 or 4+2 erasure code for now so the overhead of storing the full uuids for fragment location is not yet important. All fragment IDs should start with the same key as the object (part) key. If the object (part) key is indexed we can quickly find to which object belong the missing fragment (e.g. for repair, see below). The hdclient backend in cloud server will return this constructed key to cloud server main code.
+
+Proposed key generation scheme:
+* <genkey> := <version>#<topology_version>#<split>#<rep_policy>#<object_key>#<rand>#<location>[#<location>]
+* <version> := Natural (so 0 or 1 to start)
+* <topology_version> := Natural (so 0 or 1 to start) - if we use indirection instead of storing raw UUIDS we must versioned the mapping
+* <split> := <n_chunk>,<split_size>
+* <n_chunk> := number of splitted parts (1 for non-splitted objects)
+* <split_size> := size of each splitted parts, except last one (see hyperdrive keys below)
+* <rep_policy> := RS,k,m or CP,n (for n-1 copies)
+* <object_key> := parent S3 object key (or a prefix) - can contain anything but ‘#’
```

Comment:
given that we accept a prefix here, how do we make the decision of truncating the key ? maximum length ?
What about the handling of `#` in the S3 object Key ? http-encoded ? other ?

Reply:
That's part of my own open questions (waiting for Arnaud or Vianney to wake up...). I wanted a character very unlikely to occur in an object name. As for the maximum length, my only constraints is the maximum key length of the hyperdrive. We can bump it but perhaps it is not worth storing the whole key.